File: SCLS-main/data_read.m
function [X, y, z,const, gamma_list, gamma_time, datasize_list] =  data_read(dataname)
switch dataname
    case {'wine_modest','wine_severe'}
        table = readtable('datasets/wine.csv');
        X = table{:, 1:11};
        y = table{:, 12};
        z = y*1;
        z(z > 10.0) = 10.0;
        z(z < 0.0) = 0.0;
        if contains(dataname,'modest')
            z(z <6) = 6;
        else
            z(z <7) = 7;
        end
        const = 0.1; % the constant for normalization
        gamma_list = linspace(1e-3,0.75,40)';% setting of ICML
%         gamma_list = 2;
        gamma_time = 0.5;
        datasize_list = linspace(150,1500,16)';
%         datasize_list = linspace(150,1500,21)';
    case {'insurance_modest','insurance_severe'}
        csvname = ['datasets/insurance.csv'];
        table = readtable(csvname);
        T = createOneHotEncoding(table, 'sex');
        T = createOneHotEncoding(T, 'smoker');
        T = createOneHotEncoding(T, 'region');
        y = T.charges;
        if contains(dataname,'modest')
            z = y - 100;
        else
            z = y - 300;
        end
        
        z(z < 0.0) = 0.0;
        z = z / 100;
        y = y / 100;
        T = removevars(T,{'charges'});
        X = T{:,:};
        const = 0.03; % the constant for normalization
%         gamma_list = linspace(1e-3,0.75,40)';% setting of ICML
        gamma_list = [1e-3];
        gamma_time = 0.5;
        datasize_list = linspace(100,1300,25)';  
    case {'building_modest','building_severe'}
        Table=xlsread('datasets/building.xlsx');
        y=Table(:,108);
        X=Table(:,1:107);
        if contains(dataname,'modest')
            z = y + 20 ;
        else
            z = y + 40 ; 
        end
        z(z < 0.0) = 0.0; 
        const = 0.01; % the constant for normalization
%         gamma_list = linspace(1e-3,0.75,40)';% setting of ICML
        gamma_list = [20];
        gamma_time = 0.5;
        datasize_list = linspace(100,300,21)'; 
        nor = const*max(abs(y));
        y = y/nor;
        z = z/nor;
        
    case {'building_modest170','building_severe170'}
        Table=xlsread('datasets/building.xlsx');
        y=Table(:,108);
        X=Table(:,1:107);
        if contains(dataname,'modest')
            z = y + 20 ;
        else
            z = y + 40 ; 
        end
        z(z < 0.0) = 0.0; 
        const = 0.01; % the constant for normalization
%         gamma_list = linspace(1e-3,0.75,40)';% setting of ICML
        gamma_list = [20];
        gamma_time = 0.5;
        datasize_list = [170]'; 
        nor = const*max(abs(y));
        y = y/nor;
        z = z/nor;        
      
 
    case{'blog_modest','blog_severe'}
        Table = csvread('datasets/blog.csv');
        y = Table(:,281);
        X = Table(:,1:280);
       if contains(dataname,'modest')
           z=y+5;
       else
           z=y+10;
       end
       z(z<0)=0;        
       const = 0.01; % the constant for normalization
%        gamma_list = linspace(1e-3,0.75,40)';% setting of ICML
       gamma_list = 200;
       gamma_time = 0.5;
       datasize_list = linspace(5000,50000,10)';  
end

File: python-impl/data_read.py
import os

import numpy as np
import pandas as pd


# Read data from database provided
# "name" as the name of the dataset
def data_read(name: str, directory="../datasets/"):
    if name in ["wine_modest", "wine_severe"]:
        data = pd.read_csv(os.path.join(directory, "wine.csv"),
                           sep=';').to_numpy()
        X, y = data[:, :-1], data[:, -1]
        z = np.clip(y, 0, 10)
        if name.find("modest") != -1:
            z = np.clip(z, 6, None)
        else:
            z = np.clip(z, 7, None)
        # Configuration parameters, can change later using
        # structure options
        const = 0.1
        gamma_list = np.linspace(1e-3, 0.75, 40)
        gamma_time = 0.5
        datasize_list = np.linspace(150, 1500, 16)

    elif name in ["insurance_modest", "insurance_severe"]:
        data = pd.read_csv(os.path.join(directory,
                                        "insurance.csv"), sep=',')
        y = data["charges"].to_numpy()
        data = data.drop("charges")
        for column in ["sex", "smoker", "region"]:
            one_hot = pd.get_dummies(data[column])
            data = data.drop(column, axis=1)
            data = data.join(one_hot)
        X = data.to_numpy()
        if name.find("modest"):
            z = y - 100
        else:
            z = y - 300
        z = np.clip(z, 0, None)
        z /= 100
        y /= 100

        # Configuration parameters
        const = 0.03
        gamma_list = [1e-3]
        gamma_time = 0.5
        datasize_list = np.linspace(100, 1300, 25)

    elif name in ["building_modest", "building_severe",
                  "building_modest170", "building_severe170"]:
        data = pd.read_excel(io=os.path.join(directory, "building.xlsx"),
                             sheet_name="Data", header=1).to_numpy()
        y = data[:, -1]
        X = data[:, :-1]
        if name.find("modest"):
            z = y + 20
        else:
            z = y + 40
        z = np.clip(z, 0, None)

        # Configuration parameters
        const = 0.01
        gamma_list = [20]
        gamma_time = 0.5
        if name in ["building_modest", "building_severe"]:
            datasize_list = np.linspace(100, 300, 21)
        else:
            datasize_list = [170]
        nor = const * np.max(np.abs(y))
        y = y / nor
        z = z / nor

    elif name in ["blog_modest", "blog_severe"]:
        data = pd.read_csv(os.path.join(directory, "blog.csv"),
                           sep=',').to_numpy()
        y = data[:, -1]
        X = data[:, :-1]
        if name.find("modest"):
            z = y + 5
        else:
            z = y + 10
        z = np.clip(z, 0, None)

        # Configuration parameters
        const = 0.01
        gamma_list = [200]
        gamma_time = 0.5
        datasize_list = np.linspace(5000, 50000, 10)

    else:
        raise ValueError("Data set name not recognized.")

    return X, y, z, const, gamma_list, gamma_time, datasize_list
